// md-editor-v3/lib/es/MdEditor.mjs
import { a as J, D as lt } from "./chunks/index.mjs";
import { s as at, a as nt, M as We } from "./chunks/index4.mjs";
import { C as rt, c as st, a as it, e as ct, b as ut, u as dt, d as mt, f as bt, g as vt, h as gt, i as ft, j as ht, k as pt, M as Tt } from "./chunks/index3.mjs";
import wt from "./ModalToolbar.mjs";
import yt from "./NormalToolbar.mjs";
import { inject as r, watch as E, nextTick as re, onMounted as _, ref as x, defineComponent as C, createVNode as l, reactive as U, toRef as Ce, Fragment as _e, computed as K, cloneVNode as ae, createApp as Ct, shallowRef as kt, onBeforeUnmount as ge } from "vue";
import { g as G, p as a, a as $t, M as oe, d as xt, b as St } from "./chunks/config.mjs";
import { linkTo as It, throttle as Et, createSmoothScroll as ze, draggingScroll as Lt } from "@vavt/util";
import { deleteLine as Nt, history as ke, defaultKeymap as At, historyKeymap as Dt, indentWithTab as Mt, undo as Rt, redo as Ft } from "@codemirror/commands";
import { markdown as Bt } from "@codemirror/lang-markdown";
import { languages as Pt } from "@codemirror/language-data";
import { EditorState as ie, EditorSelection as W, Compartment as q, StateEffect as ne, StateField as qe, RangeSetBuilder as Ht } from "@codemirror/state";
import { placeholder as Ot, EditorView as R, showTooltip as Vt, Decoration as Wt, WidgetType as _t, keymap as zt, drawSelection as qt } from "@codemirror/view";
import { b as g, E as j, R as y, C as Ut, U as fe, a as Ue, c as Ke, O as Ge, d as Kt, e as Gt, T as jt, S as Yt, G as $e } from "./chunks/event-bus.mjs";
import { indentUnit as Zt, syntaxHighlighting as je, HighlightStyle as Ye } from "@codemirror/language";
import { autocompletion as Xt } from "@codemirror/autocomplete";
import { a as Jt } from "./chunks/dom.mjs";
import { I as $, M as Qt } from "./chunks/index2.mjs";
import { b as eo } from "./chunks/index5.mjs";
import { tags as d } from "@lezer/highlight";
import { searchKeymap as to } from "@codemirror/search";
import { g as xe } from "./chunks/vue-tsx.mjs";
const oo = (e, t, o) => {
  const n = r("editorId"), i = r("setting");
  let s = () => {
  }, c = () => {
  };
  const u = () => {
    s();
    const m = o.value?.view.contentDOM.getRootNode(), v = m?.querySelector(
      `#${n} .cm-scroller`
    ), b = m?.querySelector(
      `[id="${n}-preview-wrapper"]`
    ), h = m?.querySelector(
      `[id="${n}-html-wrapper"]`
    );
    (b || h) && ([c, s] = (b ? at : nt)(
      v,
      b || h,
      o.value
    ), e.scrollAuto && c());
  };
  E([t, i], () => {
    re(u);
  }), E(
    () => e.scrollAuto,
    (m) => {
      m ? c() : s();
    }
  ), E(
    () => i.value.previewOnly,
    (m) => {
      m ? s() : c();
    }
  ), _(u);
}, ce = async (e, t, o) => {
  if (/^h[1-6]$/.test(e))
    return lo(e, t);
  if (e === "prettier")
    return await ao(t, o);
  switch (e) {
    case "bold":
    case "underline":
    case "italic":
    case "strikeThrough":
    case "sub":
    case "sup":
    case "codeRow":
    case "katexInline":
    case "katexBlock":
      return ro(e, t);
    case "quote":
    case "orderedList":
    case "unorderedList":
    case "task":
      return io(e, t);
    case "code":
      return co(o, t);
    case "table":
      return bo(o);
    case "link": {
      const n = t.getSelectedText(), { desc: i = n, url: s = "" } = o, c = `[${i}](${s})`;
      return {
        text: c,
        options: {
          select: s === "",
          deviationStart: c.length - s.length - 1,
          deviationEnd: -1
        }
      };
    }
    case "image":
      return mo(o, t);
    case "flow":
    case "sequence":
    case "gantt":
    case "class":
    case "state":
    case "pie":
    case "relationship":
    case "journey":
      return uo(e);
    case "universal":
      return vo(t.getSelectedText(), o);
    default:
      return { text: "", options: {} };
  }
}, lo = (e, t) => {
  const o = e.slice(1), n = "#".repeat(Number(o)), [i, s, c] = he(t, {
    wholeLine: !0
  });
  return {
    text: `${n} ${i}`,
    options: { deviationStart: n.length + 1, replaceStart: s, replaceEnd: c }
  };
}, ao = async (e, t) => {
  const o = window.prettier || G.editorExtensions.prettier?.prettierInstance, n = [
    window.prettierPlugins?.markdown || G.editorExtensions.prettier?.parserMarkdownInstance
  ];
  return !o || !n[0] ? (g.emit(t.editorId, j, {
    name: "prettier",
    message: "prettier is undefined"
  }), {
    text: e.getValue(),
    options: { select: !1, replaceAll: !0 }
  }) : {
    text: await o.format(e.getValue(), {
      parser: "markdown",
      plugins: n
    }),
    options: { select: !1, replaceAll: !0 }
  };
}, no = {
  bold: ["**", "**", 2, -2],
  underline: ["<u>", "</u>", 3, -4],
  italic: ["*", "*", 1, -1],
  strikeThrough: ["~~", "~~", 2, -2],
  sub: ["~", "~", 1, -1],
  sup: ["^", "^", 1, -1],
  codeRow: ["`", "`", 1, -1],
  katexInline: ["$", "$", 1, -1],
  katexBlock: [`
$$
`, `
$$
`, 4, -4]
}, ro = (e, t) => {
  const o = t.getSelectedText(), [n, i, s, c] = no[e];
  return {
    text: `${n}${o}${i}`,
    options: {
      deviationStart: s,
      deviationEnd: c
      // replaceStart, replaceEnd
    }
  };
}, so = {
  quote: "> ",
  unorderedList: "- ",
  orderedList: 1,
  task: "- [ ] "
}, io = (e, t) => {
  const [o, n, i] = he(t, {
    wholeLine: !0
  }), s = o.split(`
`), c = so[e], u = e === "orderedList" ? s.map((b, h) => `${c + h}. ${b}`) : s.map((b) => `${c}${b}`), m = e === "orderedList" ? "1. " : c.toString(), v = s.length === 1 ? m.length : 0;
  return {
    text: u.join(`
`),
    options: { deviationStart: v, replaceStart: n, replaceEnd: i }
  };
}, co = (e, t) => {
  const [o, n, i] = he(t), s = e.mode || "language", c = `
\`\`\`${s}
${e.text || o || ""}
\`\`\`
`;
  return {
    text: c,
    options: {
      deviationStart: 4,
      deviationEnd: 4 + s.length - c.length,
      replaceStart: n,
      replaceEnd: i
    }
  };
}, uo = (e) => ({
  text: `
\`\`\`mermaid
${{
    flow: `flowchart TD 
  Start --> Stop`,
    sequence: `sequenceDiagram
  A->>B: hello!
  B-->>A: hi!`,
    gantt: `gantt
title Gantt Chart
dateFormat  YYYY-MM-DD`,
    class: `classDiagram
  class Animal`,
    state: `stateDiagram-v2
  s1 --> s2`,
    pie: `pie
  "Dogs" : 386
  "Cats" : 85
  "Rats" : 15`,
    relationship: `erDiagram
  CAR ||--o{ NAMED-DRIVER : allows`,
    journey: `journey
  title My Journey`,
    ...G.editorConfig.mermaidTemplate
  }[e]}
\`\`\`
`,
  options: { deviationStart: 12, deviationEnd: -5 }
}), mo = (e, t) => {
  const o = t.getSelectedText(), { desc: n = o, url: i = "", urls: s } = e;
  let c = "";
  const u = i === "" && (!s || s instanceof Array && s.length === 0);
  return s instanceof Array ? c = s.reduce((m, v) => {
    const {
      url: b = "",
      alt: h = "",
      title: f = ""
    } = typeof v == "object" ? v : { url: v };
    return m + `![${h}](${b}${f ? " '" + f + "'" : ""})
`;
  }, "") : c = `![${n}](${i})
`, {
    text: c,
    options: {
      select: i === "",
      deviationStart: u ? c.length - i.length - 2 : c.length,
      deviationEnd: u ? -2 : 0
    }
  };
}, bo = (e) => {
  const { selectedShape: t = { x: 1, y: 1 } } = e, { x: o, y: n } = t;
  let i = `
| Column`;
  for (let s = 0; s <= n; s++)
    i += " |";
  i += `
|`;
  for (let s = 0; s <= n; s++)
    i += " - |";
  for (let s = 0; s <= o; s++) {
    i += `
|`;
    for (let c = 0; c <= n; c++)
      i += " |";
  }
  return i += `
`, { text: i, options: { deviationStart: 3, deviationEnd: 10 - i.length } };
}, vo = (e, t) => {
  const { generate: o } = t, n = o(e);
  return {
    text: n.targetValue,
    options: {
      select: n.select ?? !0,
      deviationStart: n.deviationStart || 0,
      deviationEnd: n.deviationEnd || 0
    }
  };
}, he = (e, t = { wholeLine: !1 }) => {
  const o = e.view.state, n = o.selection.main;
  if (n.empty) {
    const i = o.doc.lineAt(n.from);
    return [o.doc.lineAt(n.from).text, i.from, i.to];
  } else if (t.wholeLine) {
    const i = o.doc.lineAt(n.from), s = o.doc.lineAt(n.to);
    return [
      o.doc.sliceString(i.from, s.to),
      i.from,
      s.to
    ];
  }
  return [
    o.doc.sliceString(n.from, n.to),
    n.from,
    n.to
  ];
}, z = (e) => {
  const t = new q();
  return (n) => (t.get(e.state) ? e.dispatch({ effects: t.reconfigure(n) }) : e.dispatch({
    effects: ne.appendConfig.of(t.of(n))
  }), !0);
};
class go {
  view;
  maxLength = Number.MAX_SAFE_INTEGER;
  // 切换tabSize的执行方法。切换时，Compartment实例需要相同
  toggleTabSize;
  togglePlaceholder;
  /**
   * 设置全部的扩展
   */
  setExtensions;
  toggleDisabled;
  toggleReadOnly;
  toggleMaxlength;
  getValue() {
    return this.view.state.doc.toString();
  }
  /**
   * 设置内容
   *
   * @param insert 待插入内容
   * @param from 插入开始位置
   * @param to 插入结束位置
   */
  setValue(t, o = 0, n = this.view.state.doc.length) {
    this.view.dispatch({
      changes: {
        from: o,
        to: n,
        insert: t
      }
    });
  }
  /**
   * 获取选中的文本
   */
  getSelectedText() {
    const { from: t, to: o } = this.view.state.selection.main;
    return this.view.state.sliceDoc(t, o);
  }
  /**
   * 使用新的内容替换选中的内容
   *
   * @param text 待替换内容
   * @param options 替换后是否选中
   */
  replaceSelectedText(t, o, n) {
    const i = {
      // 是否选中
      select: !0,
      // 选中时，开始位置的偏移量
      deviationStart: 0,
      // 结束的偏移量
      deviationEnd: 0,
      // 直接替换所有文本
      replaceAll: !1,
      // 替换旧文本的开始位置
      replaceStart: -1,
      // 替换旧文本的结束位置
      replaceEnd: -1,
      ...o
    };
    try {
      if (i.replaceAll) {
        if (this.setValue(t), t.length > this.maxLength)
          throw new Error("The input text is too long");
        return;
      }
      if (this.view.state.doc.length - this.getSelectedText().length + t.length > this.maxLength)
        throw new Error("The input text is too long");
      const { from: s } = this.view.state.selection.main;
      i.replaceStart !== -1 ? this.view.dispatch({
        changes: {
          from: i.replaceStart,
          to: i.replaceEnd,
          insert: t
        }
      }) : this.view.dispatch(this.view.state.replaceSelection(t)), i.select && this.view.dispatch({
        selection: {
          anchor: i.replaceStart === -1 ? s + i.deviationStart : i.replaceStart + i.deviationStart,
          head: i.replaceStart === -1 ? s + t.length + i.deviationEnd : i.replaceStart + t.length + i.deviationEnd
        }
      }), this.view.focus();
    } catch (s) {
      if (s.message === "The input text is too long")
        g.emit(n, j, {
          name: "overlength",
          message: s.message,
          data: t
        });
      else
        throw s;
    }
  }
  constructor(t) {
    this.view = t, this.toggleTabSize = z(this.view), this.togglePlaceholder = z(this.view), this.setExtensions = z(this.view), this.toggleDisabled = z(this.view), this.toggleReadOnly = z(this.view), this.toggleMaxlength = z(this.view);
  }
  /**
   * 设置tabSize
   *
   * @param tabSize 需要切换的大小
   */
  setTabSize(t) {
    this.toggleTabSize([
      ie.tabSize.of(t),
      Zt.of(" ".repeat(t))
    ]);
  }
  /**
   * 设置placeholder
   *
   * @param t 目标内容
   */
  setPlaceholder(t) {
    this.togglePlaceholder(Ot(t));
  }
  focus(t) {
    if (this.view.focus(), !t)
      return;
    let o = 0, n = 0, i = 0;
    switch (t) {
      case "start":
        break;
      case "end": {
        o = n = i = this.getValue().length;
        break;
      }
      default:
        o = t.rangeAnchor || t.cursorPos, n = t.rangeHead || t.cursorPos, i = t.cursorPos;
    }
    this.view.dispatch({
      scrollIntoView: !0,
      selection: W.create(
        [W.range(o, n), W.cursor(i)],
        1
      )
    });
  }
  setDisabled(t) {
    this.toggleDisabled([R.editable.of(!t)]);
  }
  setReadOnly(t) {
    this.toggleReadOnly([ie.readOnly.of(t)]);
  }
  setMaxLength(t) {
    this.maxLength = t, this.toggleMaxlength([
      ie.changeFilter.of((o) => o.newDoc.length <= t)
    ]);
  }
}
const fo = (e, t) => {
  if (e === t) return !0;
  if (e.length !== t.length) return !1;
  for (let o = 0; o < e.length; o++)
    if (e[o] !== t[o]) return !1;
  return !0;
}, ho = (e, t) => {
  const o = x(t.value);
  E([t], () => {
    (!o.value || !fo(o.value, t.value)) && (o.value = t.value, e());
  });
}, le = (e, t, o, n, i) => (s, c, u, m) => {
  const v = `${e}${t}${o}${n}`, b = u + c.label.length + (i === "title" ? o.length : 0);
  s.dispatch({
    changes: { from: u, to: m, insert: v },
    selection: W.create(
      [
        W.range(
          u + c.label.length + (i === "title" ? 1 : -t.length),
          b
        ),
        W.cursor(b)
      ],
      1
    )
  }), s.focus();
}, Se = (e) => (t, o, n, i) => {
  const s = e.slice(i - n);
  t.dispatch(t.state.replaceSelection(`${s} `));
}, Ie = (e) => {
  const t = (o) => {
    const n = o.matchBefore(
      /^#+|^-\s*\[*\s*\]*|`+|\[|!\[*|^\|\s?\|?|\$\$?|!+\s*\w*/
    );
    return n === null || n.from == n.to && o.explicit ? null : {
      from: n.from,
      options: [
        // 标题
        ...["h2", "h3", "h4", "h5", "h6"].map((i, s) => {
          const c = new Array(s + 2).fill("#").join("");
          return {
            label: c,
            type: "text",
            apply: Se(c)
          };
        }),
        // 任务列表
        ...["unchecked", "checked"].map((i) => {
          const s = i === "checked" ? "- [x]" : "- [ ]";
          return {
            label: s,
            type: "text",
            apply: Se(s)
          };
        }),
        // 代码
        ...[
          ["`", ""],
          ["```", "language"],
          ["```mermaid\n", ""],
          ["```echarts\n", ""]
        ].map((i) => ({
          label: `${i[0]}${i[1]}`,
          type: "text",
          apply: le(i[0], i[1], "", i[0] === "`" ? "`" : "\n```", "type")
        })),
        // 链接
        {
          label: "[]()",
          type: "text"
        },
        {
          label: "![]()",
          type: "text"
        },
        // 表格
        {
          label: "| |",
          type: "text",
          detail: "table",
          apply: `| col | col | col |
| - | - | - |
| content | content | content |
| content | content | content |`
        },
        // 公式
        {
          label: "$",
          type: "text",
          apply: le("$", "", "", "$", "type")
        },
        {
          label: "$$",
          type: "text",
          apply: le("$$", "", `
`, `
$$`, "title")
        },
        // 那啥？
        ...[
          "note",
          "abstract",
          "info",
          "tip",
          "success",
          "question",
          "warning",
          "failure",
          "danger",
          "bug",
          "example",
          "quote",
          "hint",
          "caution",
          "error",
          "attention"
        ].map((i) => ({
          label: `!!! ${i}`,
          type: "text",
          apply: le("!!!", ` ${i}`, " Title", `

!!!`, "title")
        }))
      ]
    };
  };
  return Xt({
    override: e ? [t, ...e] : [t]
  });
}, po = /* @__PURE__ */ C({
  name: `${a}-divider`,
  setup() {
    return () => l("div", {
      class: `${a}-divider`
    }, null);
  }
}), To = /* @__PURE__ */ C({
  name: "ToolbarBold",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.bold,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "bold");
      },
      type: "button"
    }, [l($, {
      name: "bold"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.bold])]);
  }
}), wo = /* @__PURE__ */ C({
  name: "ToolbarCatalog",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = r("catalogVisible");
    return () => l("button", {
      class: [`${a}-toolbar-item`, i.value && `${a}-toolbar-active`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.catalog,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, Ut);
      },
      key: "bar-catalog",
      type: "button"
    }, [l($, {
      name: "catalog"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.catalog])]);
  }
}), yo = /* @__PURE__ */ C({
  name: "ToolbarCode",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.code,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "code");
      },
      type: "button"
    }, [l($, {
      name: "code"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.code])]);
  }
}), Co = /* @__PURE__ */ C({
  name: "ToolbarCodeRow",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.codeRow,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "codeRow");
      },
      type: "button"
    }, [l($, {
      name: "code-row"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.codeRow])]);
  }
}), ko = /* @__PURE__ */ C({
  name: "ToolbarFullscreen",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), {
      fullscreenHandler: i
    } = tl();
    return () => l("button", {
      class: [`${a}-toolbar-item`, n.value.fullscreen && `${a}-toolbar-active`, t?.value && `${a}-disabled`],
      title: e.value.toolbarTips?.fullscreen,
      disabled: t?.value,
      onClick: () => {
        i();
      },
      type: "button"
    }, [l($, {
      name: n.value.fullscreen ? "fullscreen-exit" : "fullscreen"
    }, null), o?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [e.value.toolbarTips?.fullscreen])]);
  }
}), $o = /* @__PURE__ */ C({
  name: "ToolbarGithub",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, t?.value && `${a}-disabled`],
      title: e.value.toolbarTips?.github,
      disabled: t?.value,
      onClick: () => {
        It("https://github.com/imzbf/md-editor-v3");
      },
      type: "button"
    }, [l($, {
      name: "github"
    }, null), o?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [e.value.toolbarTips?.github])]);
  }
}), xo = /* @__PURE__ */ C({
  name: "ToolbarHtmlPreview",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => l("button", {
      class: [`${a}-toolbar-item`, n.value.htmlPreview && `${a}-toolbar-active`, t?.value && `${a}-disabled`],
      title: e.value.toolbarTips?.htmlPreview,
      disabled: t?.value,
      onClick: () => {
        i("htmlPreview");
      },
      type: "button"
    }, [l($, {
      name: "preview-html"
    }, null), o?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [e.value.toolbarTips?.htmlPreview])]);
  }
}), So = /* @__PURE__ */ C({
  name: "ToolbarImage",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.image,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "image");
      },
      type: "button"
    }, [l($, {
      name: "image"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.image])]);
  }
}), Io = {
  visible: {
    type: Boolean,
    default: !1
  },
  onCancel: {
    type: Function,
    default: () => {
    }
  },
  onOk: {
    type: Function,
    default: () => {
    }
  }
}, Eo = /* @__PURE__ */ C({
  name: `${a}-modal-clip`,
  props: Io,
  setup(e) {
    const t = r("usedLanguageText"), o = r("editorId"), n = r("rootRef");
    let i = G.editorExtensions.cropper.instance;
    const s = x(), c = x(), u = x(), m = U({
      cropperInited: !1,
      imgSelected: !1,
      imgSrc: "",
      // 是否全屏
      isFullscreen: !1
    });
    let v = null;
    E(() => e.visible, () => {
      e.visible && !m.cropperInited && (i = i || window.Cropper, s.value.onchange = () => {
        if (!i) {
          g.emit(o, j, {
            name: "Cropper",
            message: "Cropper is undefined"
          });
          return;
        }
        const h = s.value.files || [];
        if (m.imgSelected = !0, h?.length > 0) {
          const f = new FileReader();
          f.onload = (p) => {
            m.imgSrc = p.target.result;
          }, f.readAsDataURL(h[0]);
        }
      });
    }), E(() => [m.imgSelected], () => {
      u.value.style = "";
    }), E([Ce(() => m.isFullscreen), Ce(() => m.imgSrc)], () => {
      m.imgSrc && re(() => {
        v?.destroy(), u.value.style = "", c.value && (v = new i(c.value, {
          viewMode: 2,
          preview: n.value.getRootNode().querySelector(`.${a}-clip-preview-target`)
          // aspectRatio: 16 / 9,
        }));
      });
    });
    const b = () => {
      v.clear(), v.destroy(), v = null, s.value.value = "", m.imgSelected = !1, m.imgSrc = "";
    };
    return () => l(Qt, {
      class: `${a}-modal-clip`,
      title: t.value.clipModalTips?.title,
      visible: e.visible,
      onClose: e.onCancel,
      showAdjust: !0,
      isFullscreen: m.isFullscreen,
      onAdjust: (h) => {
        m.isFullscreen = h;
      },
      width: "668px",
      height: "421px"
    }, {
      default: () => [l("div", {
        class: `${a}-form-item ${a}-clip`
      }, [l("div", {
        class: `${a}-clip-main`
      }, [m.imgSelected ? l("div", {
        class: `${a}-clip-cropper`
      }, [l("img", {
        src: m.imgSrc,
        ref: c,
        style: {
          display: "none"
        },
        alt: ""
      }, null), l("div", {
        class: `${a}-clip-delete`,
        onClick: b
      }, [l($, {
        name: "delete"
      }, null)])]) : l("div", {
        class: `${a}-clip-upload`,
        onClick: () => {
          s.value.click();
        },
        role: "button",
        tabindex: "0",
        "aria-label": t.value.imgTitleItem?.upload
      }, [l($, {
        name: "upload"
      }, null)])]), l("div", {
        class: `${a}-clip-preview`
      }, [l("div", {
        class: `${a}-clip-preview-target`,
        ref: u
      }, null)])]), l("div", {
        class: `${a}-form-item`
      }, [l("button", {
        class: `${a}-btn`,
        type: "button",
        onClick: () => {
          if (v) {
            const h = v.getCroppedCanvas();
            g.emit(o, fe, [eo(h.toDataURL("image/png"))], e.onOk), b();
          }
        }
      }, [t.value.clipModalTips?.buttonUpload || t.value.linkModalTips?.buttonOK])]), l("input", {
        ref: s,
        accept: "image/*",
        type: "file",
        multiple: !1,
        style: {
          display: "none"
        },
        "aria-hidden": "true"
      }, null)]
    });
  }
}), Lo = {
  clipVisible: {
    type: Boolean,
    default: !1
  },
  onCancel: {
    type: Function,
    default: () => {
    }
  },
  onOk: {
    type: Function,
    default: () => {
    }
  }
}, No = /* @__PURE__ */ C({
  name: `${a}-modals`,
  props: Lo,
  setup(e) {
    return () => l(Eo, {
      visible: e.clipVisible,
      onOk: e.onOk,
      onCancel: e.onCancel
    }, null);
  }
}), Ao = /* @__PURE__ */ C({
  name: "ToolbarImageDropdown",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = x(!1), c = x(!1), u = x(), m = () => {
      g.emit(e, fe, Array.from(u.value.files || [])), u.value.value = "";
    }, v = (b, h) => {
      o?.value || g.emit(e, y, b, h);
    };
    return _(() => {
      u.value.addEventListener("change", m);
    }), () => l(_e, null, [l("label", {
      for: `${i}_label`,
      style: {
        display: "none"
      },
      "aria-label": t.value.imgTitleItem?.upload
    }, null), l("input", {
      id: `${i}_label`,
      ref: u,
      accept: "image/*",
      type: "file",
      multiple: !0,
      style: {
        display: "none"
      }
    }, null), l(J, {
      relative: `#${i}`,
      visible: s.value,
      onChange: (b) => {
        s.value = b;
      },
      disabled: o?.value,
      overlay: l("ul", {
        class: `${a}-menu`,
        onClick: () => {
          s.value = !1;
        },
        role: "menu"
      }, [l("li", {
        class: `${a}-menu-item ${a}-menu-item-image`,
        onClick: () => {
          v("image");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.imgTitleItem?.link]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-image`,
        onClick: () => {
          u.value.click();
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.imgTitleItem?.upload]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-image`,
        onClick: () => {
          c.value = !0;
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.imgTitleItem?.clip2upload])])
    }, {
      default: () => [l("button", {
        class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
        title: t.value.toolbarTips?.image,
        disabled: o?.value,
        type: "button"
      }, [l($, {
        name: "image"
      }, null), n?.value && l("div", {
        class: `${a}-toolbar-item-name`
      }, [t.value.toolbarTips?.image])])]
    }), l(No, {
      clipVisible: c.value,
      onCancel: () => {
        c.value = !1;
      },
      onOk: (b) => {
        b && v("image", {
          desc: b.desc,
          url: b.url,
          transform: !0
        }), c.value = !1;
      }
    }, null)]);
  }
}), Do = /* @__PURE__ */ C({
  name: "ToolbarItalic",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.italic,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "italic");
      },
      type: "button"
    }, [l($, {
      name: "italic"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.italic])]);
  }
}), Mo = /* @__PURE__ */ C({
  name: "ToolbarKatex",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = x(!1), c = (u) => {
      o?.value || g.emit(e, y, u);
    };
    return () => l(J, {
      relative: `#${i}`,
      visible: s.value,
      onChange: (u) => {
        s.value = u;
      },
      disabled: o?.value,
      overlay: l("ul", {
        class: `${a}-menu`,
        onClick: () => {
          s.value = !1;
        },
        role: "menu"
      }, [l("li", {
        class: `${a}-menu-item ${a}-menu-item-katex`,
        onClick: () => {
          c("katexInline");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.katex?.inline]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-katex`,
        onClick: () => {
          c("katexBlock");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.katex?.block])]),
      key: "bar-katex"
    }, {
      default: () => [l("button", {
        class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
        title: t.value.toolbarTips?.katex,
        disabled: o?.value,
        type: "button"
      }, [l($, {
        name: "formula"
      }, null), n?.value && l("div", {
        class: `${a}-toolbar-item-name`
      }, [t.value.toolbarTips?.katex])])]
    });
  }
}), Ro = /* @__PURE__ */ C({
  name: "ToolbarLink",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.link,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "link");
      },
      type: "button"
    }, [l($, {
      name: "link"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.link])]);
  }
}), Fo = /* @__PURE__ */ C({
  name: "ToolbarMermaid",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = x(!1), c = (u) => {
      o?.value || g.emit(e, y, u);
    };
    return () => l(J, {
      relative: `#${i}`,
      visible: s.value,
      onChange: (u) => {
        s.value = u;
      },
      disabled: o?.value,
      overlay: l("ul", {
        class: `${a}-menu`,
        onClick: () => {
          s.value = !1;
        },
        role: "menu"
      }, [l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("flow");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.flow]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("sequence");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.sequence]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("gantt");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.gantt]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("class");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.class]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("state");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.state]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("pie");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.pie]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("relationship");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.relationship]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-mermaid`,
        onClick: () => {
          c("journey");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.mermaid?.journey])]),
      key: "bar-mermaid"
    }, {
      default: () => [l("button", {
        class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
        title: t.value.toolbarTips?.mermaid,
        disabled: o?.value,
        type: "button"
      }, [l($, {
        name: "mermaid"
      }, null), n?.value && l("div", {
        class: `${a}-toolbar-item-name`
      }, [t.value.toolbarTips?.mermaid])])]
    });
  }
}), Bo = /* @__PURE__ */ C({
  name: "ToolbarNext",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.next,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, Ue);
      },
      type: "button"
    }, [l($, {
      name: "next"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.next])]);
  }
}), Po = /* @__PURE__ */ C({
  name: "ToolbarOrderedList",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.orderedList,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "orderedList");
      },
      type: "button"
    }, [l($, {
      name: "ordered-list"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.orderedList])]);
  }
}), Ho = /* @__PURE__ */ C({
  name: "ToolbarPageFullscreen",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => l("button", {
      class: [`${a}-toolbar-item`, n.value.pageFullscreen && `${a}-toolbar-active`, t?.value && `${a}-disabled`],
      title: e.value.toolbarTips?.pageFullscreen,
      disabled: t?.value,
      onClick: () => {
        i("pageFullscreen");
      },
      type: "button"
    }, [l($, {
      name: n.value.pageFullscreen ? "minimize" : "maximize"
    }, null), o?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [e.value.toolbarTips?.pageFullscreen])]);
  }
}), Oo = /* @__PURE__ */ C({
  name: "ToolbarPrettier",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.prettier,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "prettier");
      },
      type: "button"
    }, [l($, {
      name: "prettier"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.prettier])]);
  }
}), Vo = /* @__PURE__ */ C({
  name: "ToolbarPreview",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => l("button", {
      class: [`${a}-toolbar-item`, n.value.preview && `${a}-toolbar-active`, t?.value && `${a}-disabled`],
      title: e.value.toolbarTips?.preview,
      disabled: t?.value,
      onClick: () => {
        i("preview");
      },
      type: "button"
    }, [l($, {
      name: "preview"
    }, null), o?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [e.value.toolbarTips?.preview])]);
  }
}), Wo = /* @__PURE__ */ C({
  name: "ToolbarPreviewOnly",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => l("button", {
      class: [`${a}-toolbar-item`, n.value.previewOnly && `${a}-toolbar-active`, t?.value && `${a}-disabled`],
      title: e.value.toolbarTips?.previewOnly,
      disabled: t?.value,
      onClick: () => {
        i("previewOnly");
      },
      type: "button"
    }, [l($, {
      name: "preview-only"
    }, null), o?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [e.value.toolbarTips?.previewOnly])]);
  }
}), _o = /* @__PURE__ */ C({
  name: "ToolbarQuote",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.quote,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "quote");
      },
      type: "button"
    }, [l($, {
      name: "quote"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.quote])]);
  }
}), zo = /* @__PURE__ */ C({
  name: "ToolbarRevoke",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.revoke,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, Ke);
      },
      type: "button"
    }, [l($, {
      name: "revoke"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.revoke])]);
  }
}), qo = /* @__PURE__ */ C({
  name: "ToolbarSave",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.save,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, Ge);
      },
      type: "button"
    }, [l($, {
      name: "save"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.save])]);
  }
}), Uo = /* @__PURE__ */ C({
  name: "ToolbarStrikeThrough",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.strikeThrough,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "strikeThrough");
      },
      type: "button"
    }, [l($, {
      name: "strike-through"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.strikeThrough])]);
  }
}), Ko = /* @__PURE__ */ C({
  name: "ToolbarSub",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.sub,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "sub");
      },
      type: "button"
    }, [l($, {
      name: "sub"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.sub])]);
  }
}), Go = /* @__PURE__ */ C({
  name: "ToolbarSup",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.sup,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "sup");
      },
      type: "button"
    }, [l($, {
      name: "sup"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.sup])]);
  }
}), jo = {
  tableShape: {
    type: Array,
    default: () => [6, 4]
  },
  onSelected: {
    type: Function,
    default: () => {
    }
  }
}, Yo = /* @__PURE__ */ C({
  name: "TableShape",
  props: jo,
  setup(e) {
    const t = U({
      x: -1,
      y: -1
    }), o = K(() => JSON.stringify(e.tableShape)), n = () => {
      const s = [...JSON.parse(o.value)];
      return (!s[2] || s[2] < s[0]) && (s[2] = s[0]), (!s[3] || s[3] < s[3]) && (s[3] = s[1]), s;
    }, i = x(n());
    return E([o], () => {
      i.value = n();
    }), () => l("div", {
      class: `${a}-table-shape`,
      onMouseleave: () => {
        i.value = n(), t.x = -1, t.y = -1;
      }
    }, [new Array(i.value[1]).fill("").map((s, c) => l("div", {
      class: `${a}-table-shape-row`,
      key: `table-shape-row-${c}`
    }, [new Array(i.value[0]).fill("").map((u, m) => l("div", {
      class: `${a}-table-shape-col`,
      key: `table-shape-col-${m}`,
      onMouseenter: () => {
        t.x = c, t.y = m, m + 1 === i.value[0] && m + 1 < i.value[2] ? i.value[0]++ : m + 2 < i.value[0] && i.value[0] > e.tableShape[0] && i.value[0]--, c + 1 === i.value[1] && c + 1 < i.value[3] ? i.value[1]++ : c + 2 < i.value[1] && i.value[1] > e.tableShape[1] && i.value[1]--;
      },
      onClick: () => {
        e.onSelected(t);
      }
    }, [l("div", {
      class: [`${a}-table-shape-col-default`, c <= t.x && m <= t.y && `${a}-table-shape-col-include`]
    }, null)]))]))]);
  }
}), Zo = /* @__PURE__ */ C({
  name: "ToolbarTable",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = r("tableShape"), s = `${e}-toolbar-wrapper`, c = x(!1);
    return () => l(J, {
      relative: `#${s}`,
      visible: c.value,
      onChange: (u) => {
        c.value = u;
      },
      disabled: o?.value,
      key: "bar-table",
      overlay: l(Yo, {
        tableShape: i.value,
        onSelected: (u) => {
          o?.value || g.emit(e, y, "table", {
            selectedShape: u
          });
        }
      }, null)
    }, {
      default: () => [l("button", {
        class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
        title: t.value.toolbarTips?.table,
        disabled: o?.value,
        type: "button"
      }, [l($, {
        name: "table"
      }, null), n?.value && l("div", {
        class: `${a}-toolbar-item-name`
      }, [t.value.toolbarTips?.table])])]
    });
  }
}), Xo = /* @__PURE__ */ C({
  name: "ToolbarTask",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.task,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "task");
      },
      type: "button"
    }, [l($, {
      name: "task"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.task])]);
  }
}), Jo = /* @__PURE__ */ C({
  name: "ToolbarTitle",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = x(!1), c = (u) => {
      o?.value || g.emit(e, y, u);
    };
    return () => l(J, {
      relative: `#${i}`,
      visible: s.value,
      onChange: (u) => {
        s.value = u;
      },
      disabled: o?.value,
      overlay: l("ul", {
        class: `${a}-menu`,
        onClick: () => {
          s.value = !1;
        },
        role: "menu"
      }, [l("li", {
        class: `${a}-menu-item ${a}-menu-item-title`,
        onClick: () => {
          c("h1");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.titleItem?.h1]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-title`,
        onClick: () => {
          c("h2");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.titleItem?.h2]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-title`,
        onClick: () => {
          c("h3");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.titleItem?.h3]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-title`,
        onClick: () => {
          c("h4");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.titleItem?.h4]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-title`,
        onClick: () => {
          c("h5");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.titleItem?.h5]), l("li", {
        class: `${a}-menu-item ${a}-menu-item-title`,
        onClick: () => {
          c("h6");
        },
        role: "menuitem",
        tabindex: "0"
      }, [t.value.titleItem?.h6])])
    }, {
      default: () => [l("button", {
        class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
        disabled: o?.value,
        title: t.value.toolbarTips?.title,
        type: "button"
      }, [l($, {
        name: "title"
      }, null), n?.value && l("div", {
        class: `${a}-toolbar-item-name`
      }, [t.value.toolbarTips?.title])])]
    });
  }
}), Qo = /* @__PURE__ */ C({
  name: "ToolbarUnderline",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.underline,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "underline");
      },
      type: "button"
    }, [l($, {
      name: "underline"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.underline])]);
  }
}), el = /* @__PURE__ */ C({
  name: "ToolbarUnorderedList",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => l("button", {
      class: [`${a}-toolbar-item`, o?.value && `${a}-disabled`],
      title: t.value.toolbarTips?.unorderedList,
      disabled: o?.value,
      onClick: () => {
        g.emit(e, y, "unorderedList");
      },
      type: "button"
    }, [l($, {
      name: "unordered-list"
    }, null), n?.value && l("div", {
      class: `${a}-toolbar-item-name`
    }, [t.value.toolbarTips?.unorderedList])]);
  }
}), tl = () => {
  const e = r("editorId"), t = r("setting"), o = r("updateSetting"), {
    editorExtensions: n,
    editorExtensionsAttrs: i
  } = G;
  let s = n.screenfull.instance;
  const c = x(!1), u = (b) => {
    if (!s) {
      g.emit(e, j, {
        name: "fullscreen",
        message: "fullscreen is undefined"
      });
      return;
    }
    s.isEnabled ? (c.value = !0, (b === void 0 ? !s.isFullscreen : b) ? s.request() : s.exit()) : console.error("browser does not support screenfull!");
  }, m = () => {
    s && s.isEnabled && s.on("change", () => {
      (c.value || t.value.fullscreen) && (c.value = !1, o("fullscreen"));
    });
  }, v = () => {
    s = window.screenfull, m();
  };
  return _(() => {
    m(), s || Jt("script", {
      ...i.screenfull?.js,
      src: n.screenfull.js,
      id: rt.screenfull,
      onload: v
    }, "screenfull");
  }), _(() => {
    g.on(e, {
      name: Kt,
      callback: u
    });
  }), {
    fullscreenHandler: u
  };
};
let ol = 0;
const Ze = () => {
  const e = r("editorId"), t = r("theme"), o = r("previewTheme"), n = r("language"), i = r("disabled"), s = r("noUploadImg"), c = r("noPrettier"), u = r("codeTheme"), m = r("showToolbarName"), v = r("setting"), b = r("defToolbars");
  return {
    barRender: (f) => {
      if ($t.includes(f))
        switch (f) {
          case "-":
            return l(po, {
              key: `bar-${ol++}`
            }, null);
          case "bold":
            return l(To, {
              key: "bar-bold"
            }, null);
          case "underline":
            return l(Qo, {
              key: "bar-unorderline"
            }, null);
          case "italic":
            return l(Do, {
              key: "bar-italic"
            }, null);
          case "strikeThrough":
            return l(Uo, {
              key: "bar-strikeThrough"
            }, null);
          case "title":
            return l(Jo, {
              key: "bar-title"
            }, null);
          case "sub":
            return l(Ko, {
              key: "bar-sub"
            }, null);
          case "sup":
            return l(Go, {
              key: "bar-sup"
            }, null);
          case "quote":
            return l(_o, {
              key: "bar-quote"
            }, null);
          case "unorderedList":
            return l(el, {
              key: "bar-unorderedList"
            }, null);
          case "orderedList":
            return l(Po, {
              key: "bar-orderedList"
            }, null);
          case "task":
            return l(Xo, {
              key: "bar-task"
            }, null);
          case "codeRow":
            return l(Co, {
              key: "bar-codeRow"
            }, null);
          case "code":
            return l(yo, {
              key: "bar-code"
            }, null);
          case "link":
            return l(Ro, {
              key: "bar-link"
            }, null);
          case "image":
            return s ? l(So, {
              key: "bar-image"
            }, null) : l(Ao, {
              key: "bar-imageDropdown"
            }, null);
          case "table":
            return l(Zo, {
              key: "bar-table"
            }, null);
          case "revoke":
            return l(zo, {
              key: "bar-revoke"
            }, null);
          case "next":
            return l(Bo, {
              key: "bar-next"
            }, null);
          case "save":
            return l(qo, {
              key: "bar-save"
            }, null);
          case "prettier":
            return !c && l(Oo, {
              key: "bar-prettier"
            }, null);
          case "pageFullscreen":
            return !v.value.fullscreen && l(Ho, {
              key: "bar-pageFullscreen"
            }, null);
          case "fullscreen":
            return l(ko, {
              key: "bar-fullscreen"
            }, null);
          case "catalog":
            return l(wo, {
              key: "bar-catalog"
            }, null);
          case "preview":
            return l(Vo, {
              key: "bar-preview"
            }, null);
          case "previewOnly":
            return l(Wo, {
              key: "bar-previewOnly"
            }, null);
          case "htmlPreview":
            return l(xo, {
              key: "bar-htmlPreview"
            }, null);
          case "github":
            return l($o, {
              key: "bar-github"
            }, null);
          case "mermaid":
            return l(Fo, {
              key: "bar-mermaid"
            }, null);
          case "katex":
            return l(Mo, {
              key: "bar-katex"
            }, null);
        }
      else if (b.value instanceof Array) {
        const p = b.value[f];
        return p ? ae(p, {
          theme: p.props?.theme || t.value,
          previewTheme: p.props?.theme || o.value,
          language: p.props?.theme || n.value,
          codeTheme: p.props?.codeTheme || u.value,
          disabled: p.props?.disabled || i.value,
          showToolbarName: p.props?.showToolbarName || m.value,
          insert(I) {
            g.emit(e, y, "universal", {
              generate: I
            });
          }
        }) : "";
      } else if (b.value?.children instanceof Array) {
        const p = b.value.children[f];
        return p ? ae(p, {
          theme: p.props?.theme || t.value,
          previewTheme: p.props?.theme || o.value,
          language: p.props?.theme || n.value,
          codeTheme: p.props?.codeTheme || u.value,
          disabled: p.props?.disabled || i.value,
          showToolbarName: p.props?.showToolbarName || m.value,
          insert(I) {
            g.emit(e, y, "universal", {
              generate: I
            });
          }
        }) : "";
      } else
        return "";
    }
  };
}, ll = /* @__PURE__ */ C({
  name: "FloatingToolbar",
  setup() {
    const e = r("floatingToolbars"), {
      barRender: t
    } = Ze();
    return () => l("div", {
      class: `${a}-floating-toolbar`
    }, [e.value.map((o) => t(o))]);
  }
}), ve = ne.define(), al = qe.define({
  create() {
    return null;
  },
  update(e, t) {
    for (const o of t.effects) o.is(ve) && (e = o.value);
    return e;
  },
  provide: (e) => Vt.from(e)
}), nl = (e) => {
  const t = (n, i) => {
    n.dispatch({
      effects: ve.of({
        pos: i,
        above: !0,
        arrow: !0,
        create: () => {
          const s = document.createElement("div"), c = document.createElement("div");
          s.appendChild(c);
          const u = Ct(ll);
          return e.privide(u), u.mount(s), { dom: s, destroy: () => u.unmount() };
        }
      })
    });
  }, o = R.updateListener.of((n) => {
    if (n.selectionSet || n.docChanged) {
      const i = n.state, s = i.selection.main;
      if (!s.empty)
        t(n.view, s.from);
      else {
        const c = s.head, u = i.doc.lineAt(c);
        /^\s*$/.test(u.text) ? t(n.view, c) : n.view.dispatch({ effects: ve.of(null) });
      }
    }
  });
  return [al, o];
}, rl = "#e5c07b", Ee = "var(--md-color)", sl = "#56b6c2", il = "#fff", Z = "#3f4a54", Le = "#2d8cf0", cl = "#2d8cf0", ul = "#3f4a54", Ne = "#d19a66", dl = "#c678dd", ml = "#f6f6f6", bl = "#ceedfa33", Ae = "var(--md-bk-color)", ue = "var(--md-bk-color)", vl = "#bad5fa", De = "#3f4a54", gl = R.theme(
  {
    "&": {
      color: Z,
      backgroundColor: Ae
    },
    ".cm-content": {
      caretColor: De
    },
    ".cm-cursor, .cm-dropCursor": { borderLeftColor: De },
    "&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": { backgroundColor: vl },
    ".cm-panels": { backgroundColor: ml, color: Z },
    ".cm-panels.cm-panels-top": { borderBottom: "1px solid var(--md-border-color)" },
    ".cm-panels.cm-panels-bottom": { borderTop: "1px solid var(--md-border-color)" },
    ".cm-searchMatch": {
      backgroundColor: "#72a1ff59",
      outline: "1px solid #457dff"
    },
    ".cm-searchMatch.cm-searchMatch-selected": {
      backgroundColor: "#6199ff2f"
    },
    ".cm-activeLine": { backgroundColor: "#ceedfa33" },
    ".cm-selectionMatch": { backgroundColor: "#aafe661a" },
    "&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket": {
      backgroundColor: "#bad0f847"
    },
    ".cm-gutters": {
      backgroundColor: Ae,
      color: Z,
      borderRight: "1px solid",
      borderColor: "var(--md-border-color)"
    },
    ".cm-activeLineGutter": {
      backgroundColor: bl
    },
    ".cm-foldPlaceholder": {
      backgroundColor: "transparent",
      border: "none",
      color: "#ddd"
    },
    ".cm-tooltip": {
      border: "1px solid var(--md-border-color)",
      backgroundColor: ue
    },
    ".cm-tooltip .cm-tooltip-arrow:before": {
      borderTopColor: "var(--md-border-color)",
      //'transparent',
      borderBottomColor: "var(--md-border-color)"
    },
    ".cm-tooltip .cm-tooltip-arrow:after": {
      borderTopColor: ue,
      borderBottomColor: ue
    },
    ".cm-tooltip-autocomplete": {
      "& > ul > li[aria-selected]": {
        color: Z
      }
    }
  }
  // { dark: true }
), fl = Ye.define([
  { tag: d.keyword, color: dl },
  { tag: [d.name, d.deleted, d.character, d.propertyName, d.macroName], color: Ee },
  { tag: [d.function(d.variableName), d.labelName], color: cl },
  { tag: [d.color, d.constant(d.name), d.standard(d.name)], color: Ne },
  { tag: [d.definition(d.name), d.separator], color: Z },
  {
    tag: [
      d.typeName,
      d.className,
      d.number,
      d.changed,
      d.annotation,
      d.modifier,
      d.self,
      d.namespace
    ],
    color: rl
  },
  {
    tag: [
      d.operator,
      d.operatorKeyword,
      d.url,
      d.escape,
      d.regexp,
      d.link,
      d.special(d.string)
    ],
    color: sl
  },
  { tag: [d.meta, d.comment], color: Le },
  { tag: d.strong, fontWeight: "bold" },
  { tag: d.emphasis, fontStyle: "italic" },
  { tag: d.strikethrough, textDecoration: "line-through" },
  { tag: d.link, color: Le, textDecoration: "underline" },
  { tag: d.heading, fontWeight: "bold", color: Ee },
  { tag: [d.atom, d.bool, d.special(d.variableName)], color: Ne },
  { tag: [d.processingInstruction, d.string, d.inserted], color: ul },
  { tag: d.invalid, color: il }
]), Me = [
  gl,
  je(fl)
], hl = "#e5c07b", Re = "var(--md-color)", pl = "#56b6c2", Tl = "#ffffff", X = "var(--md-color)", Fe = "#e5c07b", wl = "#e5c07b", yl = "var(--md-color)", Be = "#d19a66", Cl = "#c678dd", kl = "#21252b", $l = "#2c313a", Pe = "var(--md-bk-color)", de = "var(--md-bk-color)", xl = "#ceedfa33", He = "#528bff", Sl = R.theme(
  {
    "&": {
      color: X,
      backgroundColor: Pe
    },
    ".cm-content": {
      caretColor: He
    },
    ".cm-cursor, .cm-dropCursor": { borderLeftColor: He },
    "&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": { backgroundColor: xl },
    ".cm-panels": { backgroundColor: kl, color: X },
    ".cm-panels.cm-panels-top": { borderBottom: "1px solid var(--md-border-color)" },
    ".cm-panels.cm-panels-bottom": { borderTop: "1px solid var(--md-border-color)" },
    ".cm-searchMatch": {
      backgroundColor: "#72a1ff59",
      outline: "1px solid #457dff"
    },
    ".cm-searchMatch.cm-searchMatch-selected": {
      backgroundColor: "#6199ff2f"
    },
    ".cm-activeLine": { backgroundColor: "#ceedfa33" },
    ".cm-selectionMatch": { backgroundColor: "#aafe661a" },
    "&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket": {
      backgroundColor: "#bad0f847"
    },
    ".cm-gutters": {
      backgroundColor: Pe,
      color: X,
      borderRight: "1px solid",
      borderColor: "var(--md-border-color)"
    },
    ".cm-activeLineGutter": {
      backgroundColor: $l
    },
    ".cm-foldPlaceholder": {
      backgroundColor: "transparent",
      border: "none",
      color: "#ddd"
    },
    ".cm-tooltip": {
      border: "1px solid var(--md-border-color)",
      backgroundColor: de
    },
    ".cm-tooltip .cm-tooltip-arrow:before": {
      borderTopColor: "var(--md-border-color)",
      //'transparent',
      borderBottomColor: "var(--md-border-color)"
    },
    ".cm-tooltip .cm-tooltip-arrow:after": {
      borderTopColor: de,
      borderBottomColor: de
    },
    ".cm-tooltip-autocomplete": {
      "& > ul > li[aria-selected]": {
        color: X
      }
    }
  },
  { dark: !0 }
), Il = Ye.define([
  { tag: d.keyword, color: Cl },
  { tag: [d.name, d.deleted, d.character, d.propertyName, d.macroName], color: Re },
  { tag: [d.function(d.variableName), d.labelName], color: wl },
  { tag: [d.color, d.constant(d.name), d.standard(d.name)], color: Be },
  { tag: [d.definition(d.name), d.separator], color: X },
  {
    tag: [
      d.typeName,
      d.className,
      d.number,
      d.changed,
      d.annotation,
      d.modifier,
      d.self,
      d.namespace
    ],
    color: hl
  },
  {
    tag: [
      d.operator,
      d.operatorKeyword,
      d.url,
      d.escape,
      d.regexp,
      d.link,
      d.special(d.string)
    ],
    color: pl
  },
  { tag: [d.meta, d.comment], color: Fe },
  { tag: d.strong, fontWeight: "bold" },
  { tag: d.emphasis, fontStyle: "italic" },
  { tag: d.strikethrough, textDecoration: "line-through" },
  { tag: d.link, color: Fe, textDecoration: "underline" },
  { tag: d.heading, fontWeight: "bold", color: Re },
  { tag: [d.atom, d.bool, d.special(d.variableName)], color: Be },
  { tag: [d.processingInstruction, d.string, d.inserted], color: yl },
  { tag: d.invalid, color: Tl }
]), Oe = [
  Sl,
  je(Il)
], El = (e, t) => {
  const o = r("editorId"), n = (s) => {
    s instanceof Promise ? s.then((c) => {
      g.emit(o, y, "universal", {
        generate() {
          return {
            targetValue: c
          };
        }
      });
    }).catch((c) => {
      console.error(c);
    }) : g.emit(o, y, "universal", {
      generate() {
        return {
          targetValue: s
        };
      }
    });
  };
  return (s) => {
    if (!s.clipboardData)
      return;
    if (s.clipboardData.files.length > 0) {
      const { files: f } = s.clipboardData;
      g.emit(
        o,
        fe,
        Array.from(f).filter((p) => /image\/.*/.test(p.type))
      ), s.preventDefault();
      return;
    }
    const c = s.clipboardData.getData("text/plain"), u = t.value?.view.state.selection.main.to || 0, m = t.value?.view.state.doc.lineAt(u).from || 0, v = t.value?.view.state.doc.sliceString(m, u) || "", b = /!\[.*\]\(\s*$/.test(v), h = /!\[.*\]\((.*)\s?.*\)/.test(c);
    if (b) {
      const f = e.transformImgUrl(c);
      n(f), s.preventDefault();
      return;
    } else if (h) {
      const f = c.match(new RegExp(`(?<=!\\[.*\\]\\()([^)\\s]+)(?=\\s?["']?.*["']?\\))`, "g"));
      f ? Promise.all(
        f.map((p) => e.transformImgUrl(p))
      ).then((p) => {
        n(
          p.reduce((T, I, A) => T.replace(f[A], I), c)
        );
      }).catch((p) => {
        console.error(p);
      }) : n(c), s.preventDefault();
      return;
    }
    if (e.autoDetectCode && s.clipboardData.types.includes("vscode-editor-data")) {
      const f = JSON.parse(s.clipboardData.getData("vscode-editor-data"));
      g.emit(o, y, "code", {
        mode: f.mode,
        text: s.clipboardData.getData("text/plain")
      }), s.preventDefault();
      return;
    }
    e.maxlength && c.length + e.modelValue.length > e.maxlength && g.emit(o, j, {
      name: "overlength",
      message: "The input text is too long",
      data: c
    });
  };
}, Ll = (e, t) => [
  {
    key: "Ctrl-b",
    mac: "Cmd-b",
    run: () => (g.emit(e, y, "bold"), !0)
  },
  {
    key: "Ctrl-d",
    mac: "Cmd-d",
    run: Nt,
    preventDefault: !0
  },
  {
    key: "Ctrl-s",
    mac: "Cmd-s",
    run: (M) => (g.emit(e, Ge, M.state.doc.toString()), !0),
    shift: () => (g.emit(e, y, "strikeThrough"), !0)
  },
  {
    key: "Ctrl-u",
    mac: "Cmd-u",
    preventDefault: !0,
    run: () => (g.emit(e, y, "underline"), !0),
    shift: () => (g.emit(e, y, "unorderedList"), !0)
  },
  {
    key: "Ctrl-i",
    mac: "Cmd-i",
    preventDefault: !0,
    run: () => (g.emit(e, y, "italic"), !0),
    shift: () => (g.emit(e, y, "image"), !0)
  },
  {
    key: "Ctrl-1",
    mac: "Cmd-1",
    run: () => (g.emit(e, y, "h1"), !0)
  },
  {
    key: "Ctrl-2",
    mac: "Cmd-2",
    run: () => (g.emit(e, y, "h2"), !0)
  },
  {
    key: "Ctrl-3",
    mac: "Cmd-3",
    run: () => (g.emit(e, y, "h3"), !0)
  },
  {
    key: "Ctrl-4",
    mac: "Cmd-4",
    run: () => (g.emit(e, y, "h4"), !0)
  },
  {
    key: "Ctrl-5",
    mac: "Cmd-5",
    run: () => (g.emit(e, y, "h5"), !0)
  },
  {
    key: "Ctrl-6",
    mac: "Cmd-6",
    run: () => (g.emit(e, y, "h6"), !0)
  },
  {
    key: "Ctrl-ArrowUp",
    mac: "Cmd-ArrowUp",
    run: () => (g.emit(e, y, "sup"), !0)
  },
  {
    key: "Ctrl-ArrowDown",
    mac: "Cmd-ArrowDown",
    run: () => (g.emit(e, y, "sub"), !0)
  },
  {
    key: "Ctrl-o",
    mac: "Cmd-o",
    run: () => (g.emit(e, y, "orderedList"), !0)
  },
  {
    key: "Ctrl-c",
    mac: "Cmd-c",
    shift: () => (g.emit(e, y, "code"), !0),
    any(M, k) {
      return (k.ctrlKey || k.metaKey) && k.altKey && k.code === "KeyC" ? (g.emit(e, y, "codeRow"), !0) : !1;
    }
  },
  {
    key: "Ctrl-l",
    mac: "Cmd-l",
    run: () => (g.emit(e, y, "link"), !0)
  },
  {
    key: "Ctrl-f",
    mac: "Cmd-f",
    shift: () => t.noPrettier ? !1 : (g.emit(e, y, "prettier"), !0)
  },
  {
    any: (M, k) => (k.ctrlKey || k.metaKey) && k.altKey && k.shiftKey && k.code === "KeyT" ? (g.emit(e, y, "table"), !0) : !1
  },
  ...to
], Nl = /[a-z][a-z0-9.+-]*:\/\/[^\s<>"'`()]+(?:\([^\s<>"'`]*\)[^\s<>"'`]*)*/i, Al = /\/\/[^\s<>"'`()]+/i, Dl = /data:[a-z]+\/[a-z0-9.+-]+(?:;base64)?,[a-z0-9+/=%]+/i, Ml = /\/(?!\/)[^\s<>"'`()]+/i, Ve = new RegExp(
  `(?<![a-z0-9.+-])(${Nl.source}|${Al.source}|${Dl.source}|${Ml.source})`,
  "gi"
), Rl = (e, t, o) => e.some((n) => n.from === t && n.to === o), Fl = (e) => {
  const t = e.shortenText || (() => "..."), o = ne.define(), n = ne.define(), i = (m, v) => {
    const b = new Ht(), h = [];
    for (let f = 1; f <= m.doc.lines; f++) {
      const p = m.doc.line(f), T = p.text;
      Ve.lastIndex = 0;
      let I;
      for (; I = Ve.exec(T); ) {
        if (!I[0]) continue;
        const A = I[0];
        if (A.length <= e.maxLength) continue;
        const L = p.from + I.index, S = L + A.length;
        if (Rl(v, L, S)) {
          h.push({ from: L, to: S });
          continue;
        }
        const N = t(A);
        b.add(
          L,
          S,
          Wt.replace({ widget: new s(N, A, L, S) })
        );
      }
    }
    return { deco: b.finish(), expanded: h };
  };
  class s extends _t {
    constructor(v, b, h, f) {
      super(), this.short = v, this.raw = b, this.from = h, this.to = f;
    }
    toDOM(v) {
      const b = document.createElement("span");
      return b.textContent = this.short, b.className = "cm-short-text", b.title = this.raw, b.style.display = "inline", b.style.textDecoration = "underline", b.addEventListener("mousedown", (h) => {
        h.preventDefault(), h.stopPropagation(), v.dispatch({
          selection: W.cursor(this.from),
          effects: o.of({
            from: this.from,
            to: this.to,
            expand: !0
          })
        }), v.focus();
      }), b.addEventListener("click", (h) => {
        h.preventDefault();
      }), b;
    }
    ignoreEvent() {
      return !1;
    }
    eq(v) {
      return this.short === v.short && this.raw === v.raw && this.from === v.from && this.to === v.to;
    }
  }
  const c = qe.define({
    create(m) {
      return i(m, []);
    },
    update(m, v) {
      let b = m.expanded;
      v.docChanged && b.length && (b = b.map(({ from: f, to: p }) => ({
        from: v.changes.mapPos(f, 1),
        to: v.changes.mapPos(p, -1)
      })).filter(({ from: f, to: p }) => f < p));
      let h = b !== m.expanded;
      for (const f of v.effects)
        f.is(o) ? f.value.expand ? b = [{ from: f.value.from, to: f.value.to }] : b = b.filter(
          ({ from: p, to: T }) => p !== f.value.from || T !== f.value.to
        ) : f.is(n) && b.length > 0 && (b = []);
      return !h && b !== m.expanded && (h = !0), v.docChanged || h ? i(v.state, b) : m;
    },
    provide: (m) => R.decorations.compute([m], (v) => v.field(m).deco)
  }), u = R.domEventHandlers({
    mousedown(m, v) {
      const b = v.state.field(c, !1);
      if (!b || b.expanded.length === 0)
        return !1;
      const h = m.target;
      if (h && v.dom.contains(h)) {
        const f = v.posAtDOM(h, 0);
        if (f != null && f !== -1 && b.expanded.some(
          ({ from: T, to: I }) => f >= T && f <= I
        ))
          return !1;
      }
      return v.dispatch({ effects: n.of(void 0) }), !1;
    }
  });
  return [c, u];
};
R.EDIT_CONTEXT = !1;
const Xe = (e) => e.extension instanceof Function ? e.extension(e.options) : e.extension, Bl = (e) => {
  const t = Xe(e);
  return e.compartment ? e.compartment.of(t) : t;
}, Pl = (e) => {
  const t = r("tabWidth"), o = r("editorId"), n = r("theme"), i = r("previewTheme"), s = r("language"), c = r(
    "usedLanguageText"
  ), u = r("disabled"), m = r("showToolbarName"), v = r("customIcon"), b = r("noUploadImg"), h = r("tableShape"), f = r("noPrettier"), p = r("codeTheme"), T = r("setting"), I = r("updateSetting"), A = r("catalogVisible"), L = r("defToolbars"), S = r("floatingToolbars"), N = r("rootRef"), M = x(), k = kt(), V = x(!1), Q = new q(), ee = new q(), se = new q(), pe = new q(), Je = new q(), Qe = Ll(o, {
    noPrettier: f
  }), Te = () => [
    ...Qe,
    ...At,
    ...Dt,
    Mt
  ], te = {
    paste: El(e, k),
    blur: e.onBlur,
    focus: e.onFocus,
    drop: e.onDrop,
    compositionstart: () => {
      V.value = !0;
    },
    compositionend: (w, D) => {
      V.value = !1, e.updateModelValue(D.state.doc.toString());
    },
    input: (w) => {
      e.onInput && e.onInput(w);
      const { data: D } = w;
      e.maxlength && e.modelValue.length + D.length > e.maxlength && g.emit(o, j, {
        name: "overlength",
        message: "The input text is too long",
        data: D
      });
    }
  }, et = nl({
    privide(w) {
      w.provide("editorId", o), w.provide("theme", n), w.provide("previewTheme", i), w.provide("language", s), w.provide("disabled", u), w.provide("noUploadImg", b), w.provide("tableShape", h), w.provide("noPrettier", f), w.provide("codeTheme", p), w.provide("showToolbarName", m), w.provide("setting", T), w.provide("updateSetting", I), w.provide("usedLanguageText", c), w.provide("catalogVisible", A), w.provide("defToolbars", L), w.provide("tabWidth", t), w.provide("customIcon", v), w.provide("floatingToolbars", S), w.provide("rootRef", N);
    }
  }), tt = [
    {
      type: "theme",
      extension: ({ theme: w }) => w.value === "light" ? Me : Oe,
      compartment: Q,
      options: {
        theme: n
      }
    },
    {
      type: "updateListener",
      extension: R.updateListener.of((w) => {
        w.docChanged && (e.onChange(w.state.doc.toString()), V.value || e.updateModelValue(w.state.doc.toString()));
      })
    },
    {
      type: "domEventHandlers",
      extension: R.domEventHandlers(te),
      compartment: pe
    },
    {
      type: "completions",
      extension: Ie(e.completions),
      compartment: ee
    },
    {
      type: "history",
      extension: ke(),
      compartment: se
    }
  ], we = G.codeMirrorExtensions(
    [
      // 横向换行
      {
        type: "lineWrapping",
        extension: R.lineWrapping
      },
      {
        type: "keymap",
        extension: zt.of(Te())
      },
      // 解决多行placeholder时，光标异常的情况
      {
        type: "drawSelection",
        extension: qt()
      },
      {
        type: "markdown",
        extension: Bt({ codeLanguages: Pt })
      },
      {
        type: "linkShortener",
        extension: (w) => Fl(w),
        options: {
          maxLength: 30
        }
      },
      {
        type: "floatingToolbar",
        extension: S.value.length > 0 ? et : [],
        compartment: Je
      }
    ],
    {
      editorId: o,
      theme: n.value,
      keyBindings: Te()
    }
  ), ot = () => [...tt, ...we].map(Bl);
  return _(() => {
    const w = new R({
      doc: e.modelValue,
      parent: M.value,
      extensions: ot()
    }), D = new go(w);
    k.value = D, setTimeout(() => {
      D.setTabSize(t), D.setDisabled(u.value), D.setReadOnly(e.readonly), e.placeholder && D.setPlaceholder(e.placeholder), typeof e.maxlength == "number" && D.setMaxLength(e.maxlength), e.autofocus && w.focus();
    }, 0), g.on(o, {
      name: Ke,
      callback() {
        Rt(w);
      }
    }), g.on(o, {
      name: Ue,
      callback() {
        Ft(w);
      }
    }), g.on(o, {
      name: y,
      async callback(B, F = {}) {
        if (B === "image" && F.transform) {
          const P = e.transformImgUrl(F.url);
          if (P instanceof Promise)
            P.then(async (O) => {
              const { text: H, options: Y } = await ce(
                B,
                k.value,
                { ...F, url: O }
              );
              k.value?.replaceSelectedText(H, Y, o);
            }).catch((O) => {
              console.error(O);
            });
          else {
            const { text: O, options: H } = await ce(B, k.value, {
              ...F,
              url: P
            });
            k.value?.replaceSelectedText(O, H, o);
          }
        } else {
          const { text: P, options: O } = await ce(
            B,
            k.value,
            F
          );
          k.value?.replaceSelectedText(P, O, o);
        }
      }
    }), g.on(o, {
      name: Gt,
      callback: Et((B) => {
        const F = { ...te }, P = Object.keys(te);
        for (const O in B) {
          const H = O;
          P.includes(H) ? F[H] = (Y, ye) => {
            B[H](Y, ye), Y.defaultPrevented || te[H](Y, ye);
          } : F[H] = B[H];
        }
        k.value?.view.dispatch({
          effects: pe.reconfigure(
            R.domEventHandlers(F)
          )
        });
      })
    }), g.on(o, {
      name: jt,
      callback: (B, F) => {
        const P = w.state.doc.line(B);
        w.dispatch(
          w.state.update({ changes: { from: P.from, to: P.to, insert: F } })
        );
      }
    }), g.on(o, {
      name: Yt,
      callback() {
        g.emit(o, $e, w);
      }
    }), g.emit(o, $e, w);
  }), E(
    n,
    () => {
      k.value?.view.dispatch({
        effects: Q.reconfigure(n.value === "light" ? Me : Oe)
      });
    },
    { deep: !0 }
  ), E(
    () => e.completions,
    () => {
      k.value?.view.dispatch({
        effects: ee.reconfigure(Ie(e.completions))
      });
    },
    { deep: !0 }
  ), E(
    () => e.modelValue,
    () => {
      k.value?.getValue() !== e.modelValue && k.value?.setValue(e.modelValue);
    }
  ), E(
    () => e.placeholder,
    () => {
      k.value?.setPlaceholder(e.placeholder);
    }
  ), E([u], () => {
    k.value?.setDisabled(u.value);
  }), E(
    () => e.readonly,
    () => {
      k.value?.setDisabled(e.readonly);
    }
  ), E(
    () => e.maxlength,
    () => {
      e.maxlength && k.value?.setMaxLength(e.maxlength);
    }
  ), ho(() => {
    const w = we.find(
      (D) => D.type === "floatingToolbar"
    );
    w?.compartment && (S.value.length > 0 ? k.value?.view.dispatch({
      effects: w.compartment.reconfigure(
        Xe(w)
      )
    }) : k.value?.view.dispatch({
      effects: w.compartment.reconfigure([])
    }));
  }, S), {
    inputWrapperRef: M,
    codeMirrorUt: k,
    resetHistory() {
      k.value?.view.dispatch({ effects: se.reconfigure([]) }), k.value?.view.dispatch({ effects: se.reconfigure(ke()) });
    }
  };
}, Hl = (e, t, o) => {
  const n = r("setting"), i = K(() => /px$/.test(`${e.inputBoxWidth}`) ? "50%" : e.inputBoxWidth), s = U({
    resizedWidth: i.value
  }), c = U({
    width: i.value
  }), u = U({
    left: i.value,
    display: "initial"
  }), m = (h) => {
    const f = t.value?.offsetWidth || 0, p = t.value?.getBoundingClientRect().x || 0;
    let T = h.x - p;
    T / f < oe ? T = f * oe : T > f - f * oe && (T = f - f * oe);
    const I = `${T / f * 100}%`;
    c.width = I, u.left = I, s.resizedWidth = I, e.oninputBoxWidthChange?.(I);
  }, v = (h) => {
    h.target === o.value && document.addEventListener("mousemove", m);
  }, b = () => {
    document.removeEventListener("mousemove", m);
  };
  return E([o], () => {
    document.removeEventListener("mousedown", v), document.removeEventListener("mouseup", b), document.addEventListener("mousedown", v), document.addEventListener("mouseup", b);
  }), _(() => {
    document.addEventListener("mousedown", v), document.addEventListener("mouseup", b);
  }), ge(() => {
    document.removeEventListener("mousedown", v), document.removeEventListener("mouseup", b);
  }), E([i], ([h]) => {
    s.resizedWidth = h, c.width = h, u.left = h;
  }), E(
    [
      () => n.value.htmlPreview,
      () => n.value.preview,
      () => n.value.previewOnly
    ],
    () => {
      n.value.previewOnly ? (c.width = "0%", u.display = "none") : !n.value.htmlPreview && !n.value.preview ? (c.width = "100%", u.display = "none") : (c.width = s.resizedWidth, u.display = "initial");
    },
    {
      immediate: !0
    }
  ), {
    inputWrapperStyle: c,
    resizeOperateStyle: u
  };
}, Ol = () => {
  const e = r("editorId"), t = x(!0), o = ze();
  return {
    onCatalogActive: (c, u) => {
      const m = document.querySelector(
        `#${e} .${a}-catalog-editor`
      );
      if (!u || !t.value || !m)
        return;
      const v = u.offsetTop - m.scrollTop;
      (v > 100 || v < 100) && o(m, u.offsetTop - 100);
    },
    onMouseenter: () => t.value = !1,
    onMouseleave: () => t.value = !0
  };
}, me = /* @__PURE__ */ C({
  name: `${xt}CustomScrollbar`,
  props: {
    id: {
      type: String,
      default: void 0
    },
    class: {
      type: [String, Array],
      default: void 0
    },
    /**
     * 滚动元素选择器，默认取 slot 根节点
     */
    scrollTarget: {
      type: String,
      default: void 0
    },
    /**
     * 是否一直显示轨道
     */
    alwaysShowTrack: {
      type: Boolean,
      default: !1
    },
    onMouseenter: {
      type: Function,
      default: () => {
      }
    },
    onMouseleave: {
      type: Function,
      default: () => {
      }
    }
  },
  setup(e, {
    slots: t
  }) {
    const o = x(null), n = x(null), i = x(null), s = x(null);
    let c = null, u = null, m = !1, v = 0, b = 0;
    const h = () => {
      if (!n.value || !o.value || !i.value || !s.value) return;
      const S = o.value.clientHeight, N = n.value.scrollHeight, M = n.value.scrollTop;
      if (N <= S) {
        i.value.style.display = "none", e.alwaysShowTrack || (s.value.style.display = "none");
        return;
      } else
        i.value.style.display = "block", s.value.style.display = "block";
      const k = S / N, V = Math.max(S * k, 20), Q = S - V, ee = Math.min(M * k, Q);
      i.value.style.height = `${V}px`, i.value.style.top = `${ee}px`;
    }, f = () => h(), p = (S) => {
      m = !0, v = S.clientY, b = n.value.scrollTop, document.body.style.userSelect = "none";
    }, T = (S) => {
      if (!m || !n.value || !o.value) return;
      const N = S.clientY - v, M = n.value.scrollHeight / o.value.clientHeight;
      n.value.scrollTop = b + N * M;
    }, I = () => {
      m = !1, document.body.style.userSelect = "";
    }, A = (S) => {
      n.value && n.value.removeEventListener("scroll", f), n.value = S, n.value ? (n.value.addEventListener("scroll", f), h()) : s.value && !e.alwaysShowTrack && (s.value.style.display = "none");
    }, L = () => {
      if (!o.value) return;
      const S = e.scrollTarget ? o.value.querySelector(e.scrollTarget) : o.value.firstElementChild;
      A(S);
    };
    return _(async () => {
      await re(), L(), c = new MutationObserver(() => {
        u && cancelAnimationFrame(u), u = requestAnimationFrame(() => {
          L();
        });
      }), c.observe(o.value, {
        childList: !0,
        subtree: !0
      }), window.addEventListener("resize", h), i.value?.addEventListener("mousedown", p), document.addEventListener("mousemove", T), document.addEventListener("mouseup", I);
    }), ge(() => {
      c?.disconnect(), n.value && n.value.removeEventListener("scroll", f), window.removeEventListener("resize", h), i.value?.removeEventListener("mousedown", p), document.removeEventListener("mousemove", T), document.removeEventListener("mouseup", I);
    }), () => l("div", {
      id: e.id,
      class: [`${a}-custom-scrollbar`, e.class],
      ref: o,
      onMouseenter: e.onMouseenter,
      onMouseleave: e.onMouseleave
    }, [t.default?.(), l("div", {
      class: `${a}-custom-scrollbar__track`,
      ref: s
    }, [l("div", {
      class: `${a}-custom-scrollbar__thumb`,
      ref: i
    }, null)])]);
  }
}), Vl = ze(), Wl = /* @__PURE__ */ C({
  name: "MDEditorContent",
  props: st,
  setup(e, t) {
    const o = r("editorId"), n = r("catalogVisible"), i = r("theme"), s = r("setting"), c = x(""), u = x(), m = x(), {
      inputWrapperRef: v,
      codeMirrorUt: b,
      resetHistory: h
    } = Pl(e), {
      inputWrapperStyle: f,
      resizeOperateStyle: p
    } = Hl(e, u, m);
    oo(e, c, b);
    const {
      onCatalogActive: T,
      onMouseenter: I,
      onMouseleave: A
    } = Ol();
    return t.expose({
      getSelectedText() {
        return b.value?.getSelectedText();
      },
      focus(L) {
        b.value?.focus(L);
      },
      resetHistory: h,
      getEditorView() {
        return b.value?.view;
      }
    }), () => l("div", {
      class: `${a}-content`
    }, [l("div", {
      class: `${a}-content-wrapper`,
      ref: u
    }, [l(me, {
      alwaysShowTrack: !0,
      scrollTarget: `#${o} .cm-scroller`,
      style: f
    }, {
      default: () => [l("div", {
        class: `${a}-input-wrapper`,
        ref: v
      }, null)]
    }), (s.value.htmlPreview || s.value.preview) && l("div", {
      class: `${a}-resize-operate`,
      style: p,
      ref: m
    }, null), l(me, {
      style: {
        flex: 1
      }
    }, {
      default: () => [l(it, {
        modelValue: e.modelValue,
        onChange: e.onChange,
        onHtmlChanged: (L) => {
          c.value = L, e.onHtmlChanged(L);
        },
        onGetCatalog: e.onGetCatalog,
        mdHeadingId: e.mdHeadingId,
        noMermaid: e.noMermaid,
        sanitize: e.sanitize,
        noKatex: e.noKatex,
        formatCopiedText: e.formatCopiedText,
        noHighlight: e.noHighlight,
        noImgZoomIn: e.noImgZoomIn,
        sanitizeMermaid: e.sanitizeMermaid,
        codeFoldable: e.codeFoldable,
        autoFoldThreshold: e.autoFoldThreshold,
        onRemount: e.onRemount
      }, null)]
    })]), n.value && l(me, {
      class: `${a}-catalog-${e.catalogLayout}`,
      onMouseenter: I,
      onMouseleave: A
    }, {
      default: () => [l(We, {
        theme: i.value,
        class: `${a}-catalog-editor`,
        editorId: o,
        mdHeadingId: e.mdHeadingId,
        key: "internal-catalog",
        scrollElementOffsetTop: 2,
        syncWith: s.value.preview ? "preview" : "editor",
        onClick: (L, S) => {
          if (!s.value.preview && S.line !== void 0) {
            L.preventDefault();
            const N = b.value?.view;
            if (N) {
              const M = N.state.doc.line(S.line + 1), k = N.lineBlockAt(M.from)?.top, V = N.scrollDOM;
              Vl(V, k);
            }
          }
        },
        catalogMaxDepth: e.catalogMaxDepth,
        onActive: T
      }, null)]
    })]);
  }
}), _l = /* @__PURE__ */ C({
  props: {
    modelValue: {
      type: String,
      default: ""
    }
  },
  setup(e) {
    const t = r("usedLanguageText");
    return () => l("div", {
      class: `${a}-footer-item`
    }, [l("label", {
      class: `${a}-footer-label`
    }, [`${t.value.footer?.markdownTotal}:`]), l("span", null, [e.modelValue?.length || 0])]);
  }
}), zl = {
  checked: {
    type: Boolean,
    default: !1
  },
  onChange: {
    type: Function,
    default: () => {
    }
  },
  disabled: {
    type: Boolean,
    default: void 0
  }
}, ql = /* @__PURE__ */ C({
  name: `${a}-checkbox`,
  props: zl,
  setup(e) {
    return () => l("div", {
      class: [`${a}-checkbox`, e.checked && `${a}-checkbox-checked`, e.disabled && `${a}-disabled`],
      onClick: () => {
        e.disabled || e.onChange(!e.checked);
      }
    }, null);
  }
}), Ul = {
  scrollAuto: {
    type: Boolean
  },
  onScrollAutoChange: {
    type: Function,
    default: () => {
    }
  }
}, Kl = /* @__PURE__ */ C({
  props: Ul,
  setup(e) {
    const t = r("usedLanguageText"), o = r("disabled");
    return () => l("div", {
      class: [`${a}-footer-item`, o?.value && `${a}-disabled`]
    }, [l("label", {
      class: `${a}-footer-label`,
      onClick: () => {
        o?.value || e.onScrollAutoChange(!e.scrollAuto);
      }
    }, [t?.value.footer?.scrollAuto]), l(ql, {
      checked: e.scrollAuto,
      onChange: e.onScrollAutoChange,
      disabled: o?.value
    }, null)]);
  }
}), Gl = {
  modelValue: {
    type: String,
    default: ""
  },
  footers: {
    type: Array,
    default: []
  },
  scrollAuto: {
    type: Boolean
  },
  noScrollAuto: {
    type: Boolean
  },
  onScrollAutoChange: {
    type: Function,
    default: () => {
    }
  },
  defFooters: {
    type: Object
  }
}, jl = /* @__PURE__ */ C({
  name: "MDEditorFooter",
  props: Gl,
  setup(e) {
    const t = r("theme"), o = r("language"), n = r("disabled"), i = K(() => {
      const c = e.footers.indexOf("="), u = c === -1 ? e.footers : e.footers.slice(0, c), m = c === -1 ? [] : e.footers.slice(c, Number.MAX_SAFE_INTEGER);
      return [u, m];
    }), s = (c) => {
      if (St.includes(c))
        switch (c) {
          case "markdownTotal":
            return l(_l, {
              modelValue: e.modelValue
            }, null);
          case "scrollSwitch":
            return !e.noScrollAuto && l(Kl, {
              scrollAuto: e.scrollAuto,
              onScrollAutoChange: e.onScrollAutoChange
            }, null);
        }
      else if (e.defFooters instanceof Array) {
        const u = e.defFooters[c];
        return u ? ae(u, {
          theme: u.props?.theme || t.value,
          language: u.props?.language || o.value,
          disabled: u.props?.disabled || n?.value
        }) : "";
      } else if (e.defFooters && e.defFooters.children instanceof Array) {
        const u = e.defFooters.children[c];
        return u ? ae(u, {
          theme: u.props?.theme || t.value,
          language: u.props?.language || o.value,
          disabled: u.props?.disabled || n?.value
        }) : "";
      } else
        return "";
    };
    return () => {
      const c = i.value[0].map((m) => s(m)), u = i.value[1].map((m) => s(m));
      return l("div", {
        class: `${a}-footer`
      }, [l("div", {
        class: `${a}-footer-left`
      }, [c]), l("div", {
        class: `${a}-footer-right`
      }, [u])]);
    };
  }
}), Yl = {
  // 工具栏选择显示
  toolbars: {
    type: Array,
    default: () => []
  },
  // 工具栏选择不显示
  toolbarsExclude: {
    type: Array,
    default: () => []
  }
}, Zl = /* @__PURE__ */ C({
  name: "MDEditorToolbar",
  props: Yl,
  setup(e) {
    const t = r("editorId"), o = r("showToolbarName"), n = `${t}-toolbar-wrapper`, i = x(), s = x(), {
      barRender: c
    } = Ze(), u = K(() => {
      const m = e.toolbars.filter((f) => !e.toolbarsExclude.includes(f)), v = m.indexOf("="), b = v === -1 ? m : m.slice(0, v + 1), h = v === -1 ? [] : m.slice(v, Number.MAX_SAFE_INTEGER);
      return [b, h];
    });
    return E(() => e.toolbars, () => {
      re(() => {
        i.value && Lt(i.value);
      });
    }, {
      immediate: !0
    }), () => {
      const m = u.value[0].map((b) => c(b)), v = u.value[1].map((b) => c(b));
      return l(_e, null, [e.toolbars.length > 0 && l("div", {
        class: `${a}-toolbar-wrapper`,
        ref: i,
        id: n
      }, [l("div", {
        class: [`${a}-toolbar`, o.value && `${a}-stn`]
      }, [l("div", {
        class: `${a}-toolbar-left`,
        ref: s
      }, [m]), l("div", {
        class: `${a}-toolbar-right`
      }, [v])])])]);
    };
  }
}), be = /* @__PURE__ */ C({
  name: "MdEditorV3",
  props: ut,
  emits: ct,
  setup(e, t) {
    const {
      noKatex: o,
      noMermaid: n,
      noHighlight: i
    } = e, s = U({
      scrollAuto: e.scrollAuto
    }), c = x(), u = x(), m = K(() => xe({
      props: e,
      ctx: t
    }, "defToolbars")), v = K(() => xe({
      props: e,
      ctx: t
    }, "defFooters")), b = dt(e), [h, f] = mt(e, t, {
      editorId: b
    }), p = bt(e, {
      editorId: b
    });
    return vt(e, t, {
      editorId: b
    }), gt(e), ft(e, t, {
      editorId: b
    }), ht(e, t, {
      editorId: b,
      catalogVisible: p,
      setting: h,
      updateSetting: f,
      codeRef: u
    }), pt(e, {
      rootRef: c,
      editorId: b,
      setting: h,
      updateSetting: f,
      catalogVisible: p,
      defToolbars: m
    }), ge(() => {
      g.clear(b);
    }), () => l("div", {
      id: b,
      class: [a, e.class, e.theme === "dark" && `${a}-dark`, h.fullscreen || h.pageFullscreen ? `${a}-fullscreen` : ""],
      style: e.style,
      ref: c
    }, [e.toolbars.length > 0 && l(Zl, {
      toolbars: e.toolbars,
      toolbarsExclude: e.toolbarsExclude
    }, null), l(Wl, {
      ref: u,
      modelValue: e.modelValue,
      mdHeadingId: e.mdHeadingId,
      noMermaid: n,
      sanitize: e.sanitize,
      placeholder: e.placeholder,
      noKatex: o,
      scrollAuto: s.scrollAuto,
      formatCopiedText: e.formatCopiedText,
      autofocus: e.autoFocus,
      readonly: e.readOnly,
      maxlength: e.maxLength,
      autoDetectCode: e.autoDetectCode,
      noHighlight: i,
      updateModelValue: (T) => {
        t.emit("update:modelValue", T);
      },
      onChange: (T) => {
        e.onChange?.(T), t.emit("onChange", T);
      },
      onHtmlChanged: (T) => {
        e.onHtmlChanged?.(T), t.emit("onHtmlChanged", T);
      },
      onGetCatalog: (T) => {
        e.onGetCatalog?.(T), t.emit("onGetCatalog", T);
      },
      onBlur: (T) => {
        e.onBlur?.(T), t.emit("onBlur", T);
      },
      onFocus: (T) => {
        e.onFocus?.(T), t.emit("onFocus", T);
      },
      onInput: (T) => {
        e.onInput?.(T), t.emit("onInput", T);
      },
      completions: e.completions,
      noImgZoomIn: e.noImgZoomIn,
      onDrop: (T) => {
        e.onDrop?.(T), t.emit("onDrop", T);
      },
      inputBoxWidth: e.inputBoxWidth,
      oninputBoxWidthChange: (T) => {
        e.oninputBoxWidthChange?.(T), t.emit("oninputBoxWidthChange", T);
      },
      sanitizeMermaid: e.sanitizeMermaid,
      transformImgUrl: e.transformImgUrl,
      codeFoldable: e.codeFoldable,
      autoFoldThreshold: e.autoFoldThreshold,
      onRemount: () => {
        e.onRemount?.(), t.emit("onRemount");
      },
      catalogLayout: e.catalogLayout,
      catalogMaxDepth: e.catalogMaxDepth,
      noEcharts: e.noEcharts
    }, null), e.footers.length > 0 && l(jl, {
      modelValue: e.modelValue,
      footers: e.footers,
      defFooters: v.value,
      noScrollAuto: !h.preview && !h.htmlPreview || h.previewOnly,
      scrollAuto: s.scrollAuto,
      onScrollAutoChange: (T) => s.scrollAuto = T
    }, null)]);
  }
});
be.install = (e) => (e.component(be.name, be), e.use(yt).use(lt).use(wt).use(We).use(Tt), e);
export {
  be as default
};
