// md-editor-v3/lib/es/ModalToolbar.mjs
import { defineComponent as s, createVNode as l, Fragment as r, isVNode as c } from "vue";
import { M as y } from "./chunks/index2.mjs";
import { p as a } from "./chunks/config.mjs";
import { g as d } from "./chunks/vue-tsx.mjs";
function g(e) {
  return typeof e == "function" || Object.prototype.toString.call(e) === "[object Object]" && !c(e);
}
const m = {
  title: {
    type: String,
    default: ""
  },
  modalTitle: {
    type: [String, Object],
    default: ""
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  width: {
    type: String,
    default: "auto"
  },
  height: {
    type: String,
    default: "auto"
  },
  // 展示在工具栏的内容，通常是个图标
  trigger: {
    type: [String, Object],
    default: void 0
  },
  onClick: {
    type: Function,
    default: void 0
  },
  onClose: {
    type: Function,
    default: void 0
  },
  /**
   * 显示全屏按钮
   */
  showAdjust: {
    type: Boolean,
    default: !1
  },
  isFullscreen: {
    type: Boolean,
    default: !1
  },
  onAdjust: {
    type: Function,
    default: void 0
  },
  class: {
    type: String,
    default: void 0
  },
  style: {
    type: [Object, String],
    default: void 0
  },
  showMask: {
    type: Boolean,
    default: !0
  },
  /**
   * ==没有意义，仅用于规避克隆组件自动嵌入insert方法时，传入的是该组件而产生的waring
   */
  insert: {
    type: Function,
    default: void 0
  },
  language: {
    type: String,
    default: void 0
  },
  theme: {
    type: String,
    default: void 0
  },
  previewTheme: {
    type: String,
    default: void 0
  },
  codeTheme: {
    type: String,
    default: void 0
  },
  disabled: {
    type: Boolean,
    default: void 0
  },
  showToolbarName: {
    type: Boolean,
    default: void 0
  }
  /**
   * ==结束
   */
}, i = /* @__PURE__ */ s({
  name: "ModalToolbar",
  props: m,
  emits: ["onClick", "onClose", "onAdjust"],
  setup(e, t) {
    return () => {
      const u = d({
        props: e,
        ctx: t
      }, "trigger"), f = d({
        props: e,
        ctx: t
      }, "modalTitle"), n = d({
        props: e,
        ctx: t
      });
      return l(r, null, [l("button", {
        class: [`${a}-toolbar-item`, e.disabled && `${a}-disabled`],
        title: e.title,
        disabled: e.disabled,
        onClick: () => {
          e.onClick?.(), t.emit("onClick");
        },
        type: "button"
      }, [u]), l(y, {
        style: e.style,
        class: e.class,
        width: e.width,
        height: e.height,
        title: f,
        visible: e.visible,
        showMask: e.showMask,
        onClose: () => {
          e.onClose?.(), t.emit("onClose");
        },
        showAdjust: e.showAdjust,
        isFullscreen: e.isFullscreen,
        onAdjust: (o) => {
          e.onAdjust?.(o), t.emit("onAdjust", o);
        }
      }, g(n) ? n : {
        default: () => [n]
      })]);
    };
  }
});
i.install = (e) => (e.component(i.name, i), e);
export {
  i as default
};
